subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c

```diff
@@ -224,6 +277,211 @@ static int validate_ak_order(void)
+		if (err) {
+			LOG_ERR("Failed to remove Bluetooth bond (err %d)", err);
+			return err;
+		}
+	} else {
+		err = fp_bond_delete(bond);
+		if (err) {
+			LOG_ERR("Failed to delete Fast Pair bond (err %d)", err);
+			return err;
+		}
+	}
+
+	LOG_DBG("Bond removed successfully");
+
+	return 0;
+}
+
+static void bonds_with_invalid_id_remove_completely(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		int err;
+		bool account_key_found = false;
+
+		if (bt_addr_le_eq(&fp_bonds[i].nv.addr, BT_ADDR_LE_ANY) || !fp_bonds[i].bonded ||
+		    (fp_bonds[i].nv.account_key_metadata_id == BOND_AK_ID_UNKNOWN)) {
+			continue;
+		}
+
+		for (size_t j = 0; j < account_key_count; j++) {
+			if (fp_bonds[i].nv.account_key_metadata_id ==
+			    ACCOUNT_KEY_METADATA_FIELD_GET(account_key_metadata[j], ID)) {
+				account_key_found = true;
+				break;
+			}
+		}
+
+		if (!account_key_found) {
+			LOG_DBG("Removing bond associated with invalid Account Key");
+			err = bond_remove_completely(&fp_bonds[i]);
+			if (err) {
+				LOG_ERR("Failed to remove bond (err %d)", err);
+			}
+		}
+	}
+}
+
+static void bonds_conn_ctx_reset(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		fp_bonds[i].conn_ctx = NULL;
+	}
+}
+
+static void bt_bonds_find(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		fp_bonds[i].bonded = bt_request_cb->is_addr_bonded(&fp_bonds[i].nv.addr);
+		if (fp_bonds[i].bonded) {
+			LOG_DBG("Loaded bond has been found in the Bluetooth bond list");
+		}
+	}
+}
+
+static void bond_duplicates_handle(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		bool duplicate_found = false;
+		int err;
+
+		if (bt_addr_le_eq(&fp_bonds[i].nv.addr, BT_ADDR_LE_ANY)) {
+			continue;
+		}
+
+		for (size_t j = i + 1; j < ARRAY_SIZE(fp_bonds); j++) {
+			if (bt_addr_le_eq(&fp_bonds[j].nv.addr, &fp_bonds[i].nv.addr)) {
+				/* Error during Fast Pair Procedure. */
+				duplicate_found = true;
+				LOG_DBG("Removing both duplicated bond entries at bootup");
+				err = fp_bond_delete(&fp_bonds[j]);
+				if (err) {
+					LOG_ERR("Failed to delete bond (err %d)", err);
+				}
+			}
+		}
+
+		if (duplicate_found) {
+			err = bond_remove_completely(&fp_bonds[i]);
+			if (err) {
+				LOG_ERR("Failed to remove bond (err %d)", err);
+			}
+		}
+	}
+}
+
+static void invalid_bonds_purge(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		int err;
+
+		if (bt_addr_le_eq(&fp_bonds[i].nv.addr, BT_ADDR_LE_ANY)) {
+			continue;
+		}
+
+		if (!fp_bonds[i].bonded) {
+			LOG_DBG("Deleting not bonded entry");
+			err = fp_bond_delete(&fp_bonds[i]);
+			if (err) {
+				LOG_ERR("Failed to delete bond (err %d)", err);
+			}
+
+			continue;
+		}
+
+		if (fp_bonds[i].nv.account_key_metadata_id == BOND_AK_ID_UNKNOWN) {
+			/* Fast Pair Procedure unfinished. */
+			err = bond_remove_completely(&fp_bonds[i]);
+			if (err) {
+				LOG_ERR("Failed to remove bond (err %d)", err);
+			}
+		}
```

Comment:
Isn't this case already covered by the bonds_with_invalid_id_remove_completely function?

Reply:
No, it only removes ids that are associated with old, unexistant AKs. It doesn't remove BOND_AK_ID_UNKNOWN ids. The `bonds_with_invalid_id_remove_completely` function is also used in `fp_storage_ak_save` where we don't want to remove bond entries with equal to BOND_AK_ID_UNKNOWN.

Reply:
ok

Reply:
Can we combine this function and `bonds_with_invalid_id_remove_completely` ?

Reply:
@MarekPieta, please resolve this comment during your review as you suggested an improvement in its scope